tensor2tensor/utils/registry.py

```diff
@@ -151,372 +81,285 @@ def default_object_name(obj):
-
-
-_OPTIMIZERS = {}
-
-
-def register_optimizer(name=None):
-  """Register an optimizer. name defaults to upper camel case of fn name."""
-
-  def default_opt_name(opt_fn):
-    return misc_utils.snakecase_to_camelcase(default_name(opt_fn))
-
-  def decorator(opt_fn, registration_name):
-    """Registers and returns optimizer_fn with registration_name or default."""
-    if registration_name is None:
-      registration_name = default_opt_name(opt_fn)
-
-    if registration_name in _OPTIMIZERS and not tf.executing_eagerly():
-      raise LookupError("Optimizer %s already registered." % registration_name)
-    args, varargs, keywords, _ = inspect.getargspec(opt_fn)
-
-    if len(args) != 2 or varargs is not None or keywords is not None:
-      raise ValueError("Optimizer registration function must take two "
-                       "arguments: learning_rate (float) and "
-                       "hparams (HParams).")
-    _OPTIMIZERS[registration_name] = opt_fn
-    return opt_fn
-
-  if callable(name):
-    opt_fn = name
-    registration_name = default_opt_name(opt_fn)
-    return decorator(opt_fn, registration_name=registration_name)
-
-  return lambda opt_fn: decorator(opt_fn, name)
-
+class Registry(object):
+  """Dict-like class for managing registrations."""
+  def __init__(
+      self, register_name, default_key_fn=default_name, validator=None,
+      on_set=None, value_transformer=_default_value_transformer):
+    """
+    Args:
+      register_name: str identifier for the given register. Used in error msgs.
+      default_key_fn (optional): function mapping value -> key for registration
+        when a key are not provided
+      validator (optional): if given, this is run before setting a given
+        (key, value) pair. Accepts (key, value) and should raise if there is a
+        problem. Overwriting existing keys is not allowed and is checked
+        separately.
+      on_set (optional): callback function accepting (key, value) pair
+        which is run after an item is successfully set.
+      value_transformer (optional): if run, `__getitem__` will return
+        value_transformer(key, registered_value).
+    """
+    self._register = {}
+    self._name = register_name
+    self._default_key_fn = default_key_fn
+    self._validator = validator
+    self._on_set = on_set
+    self._value_transformer = value_transformer
+
+  def default_key(self, key):
+    return self._default_key_fn(key)
+
+  @property
+  def name(self):
+    return self._name
+
+  def validate(self, key, value):
+    if self._validator is not None:
+      self._validator(key, value)
+
+  def __setitem__(self, key, value):
+    if key in self:
+      raise KeyError("key %s already registered in registry %s"
+                     % (key, self._name))
+    self.validate(key, value)
+    self._register[key] = value
+    callback = self._on_set
+    if callback is not None:
+      callback(key, value)
+
+  def register(self, key=None):
```

Comment:
add method docstrings here and in any other method that's > a few lines.
at least a 1-liner.
if > 1 line docstring, make sure it has an Args: section and a Returns: section